# MNIST

In [1]:
import numpy as np
from sklearn.datasets import fetch_mldata

In [2]:
mnist = fetch_mldata("MNIST original", data_home="D:\\datasets")

In [3]:
mnist

{'COL_NAMES': ['label', 'data'],
 'DESCR': 'mldata.org dataset: mnist-original',
 'data': array([[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]], dtype=uint8),
 'target': array([0., 0., 0., ..., 9., 9., 9.])}

In [4]:
X,y = mnist['data'], mnist['target']

In [5]:
X.shape

(70000, 784)

In [6]:
X_train = np.array(X[:60000],dtype=float)
y_train = np.array(y[:60000],dtype=float)
X_test = np.array(X[60000:],dtype=float)
y_test = np.array(y[60000:],dtype=float)

In [7]:
X_train.shape

(60000, 784)

In [8]:
y_test.shape

(10000,)

## 使用kNN

In [12]:
from sklearn.neighbors import KNeighborsClassifier

knn_clf = KNeighborsClassifier()
#%time  knn_clf.fit(X_train, y_train) #要半分钟

In [18]:
#%time knn_clf.score(X_test, y_test) #要十几分钟
#准确率大概 0.9687999999999

### 这里为什么不需要进行均值方差归一化？
#### 均值方差归一化的作用是将不同尺度的数据放在同一尺度下
#### 这里的784个维度都是亮度，本身就在同一尺度下了，不用归一化

# PCA 降维

In [13]:
from sklearn.decomposition import PCA

In [14]:
pca = PCA(0.9)
pca.fit(X_train)
X_train_reduction = pca.transform(X_train)

In [16]:
X_train_reduction.shape

(60000, 87)

In [17]:
knn_clf = KNeighborsClassifier()
%time knn_clf.fit(X_train_reduction, y_train)

Wall time: 521 ms


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')

In [19]:
X_test_reduction = pca.transform(X_test)

In [ ]:
knn_clf.score(X_test_reduction, y_test) # 大约耗时2分钟

### 降低维度 部分信息丢失 为什么准确率更高？
#### 降维过程中可以降噪